### PR TITLE
Load test fixes

### PR DESCRIPTION
**Pull request by akallabeth**

### CODE REVIEW

**Review by mfleisz — COMMENTED** [review]
Two more comments:

- `freerdp_settings_get_bool(settings, FreeRDP_DeactivateClientDecoding)` is called quite frequently (also in recv_primary_order and recv_secondary_order). Maybe it would make sense to cache the defaultReturn in the context or somewhere else?
- I saw two changes in the proxy  code where `WaitForMultipleObjects` calls were changed from `INFINITE` to 1000. One of the changes says `/* Do periodic polling to avoid client hang */`. Doesn't that suggest that there is some sort of synchronization issue here and we just "hide" it with the timeout?

File: channels/rdpsnd/client/rdpsnd_main.c [hunk 1]
```diff
@@ -1346,7 +1375,7 @@ static UINT rdpsnd_virtual_channel_event_initialized(rdpsndPlugin* rdpsnd)
 	return CHANNEL_RC_OK;
 }
 
-static void rdpsnd_virtual_channel_event_terminated(rdpsndPlugin* rdpsnd)
+void rdpsnd_virtual_channel_event_terminated(rdpsndPlugin* rdpsnd)
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply channels/rdpsnd/client/rdpsnd_main.c:1378]: Still `static`, it is declared at the beginning of the file.

File: winpr/libwinpr/thread/thread.c [hunk 2]
```diff
@@ -562,6 +593,7 @@ HANDLE CreateRemoteThread(HANDLE hProcess, LPSECURITY_ATTRIBUTES lpThreadAttribu
 
 VOID ExitThread(DWORD dwExitCode)
 {
+#if 0
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply winpr/libwinpr/thread/thread.c]: replaced with proper define

File: server/proxy/pf_server.c [hunk 3]
```diff
@@ -479,28 +495,35 @@ static DWORD WINAPI pf_server_handle_peer(LPVOID arg)
...
+		HANDLE ChannelEvent = INVALID_HANDLE_VALUE;
+		DWORD eventCount = 0;
 		{
 			WINPR_ASSERT(client->GetEventHandles);
-			tmp = client->GetEventHandles(client, &eventHandles[eventCount], 32 - eventCount);
+			tmp = client->GetEventHandles(client, &eventHandles[eventCount],
+			                              ARRAYSIZE(eventHandles) - eventCount);
 
-			if (tmp == 0)
+			if ((tmp == 0) || (tmp >= ARRAYSIZE(eventHandles) - 2))
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply server/proxy/pf_server.c]: done

File: server/proxy/pf_client.c [hunk 4]
```diff
@@ -717,13 +725,14 @@ static DWORD WINAPI pf_client_thread_proc(pClientContext* pc)
...
 	}
+	handles[nCount++] = Queue_Event(pc->cached_server_channel_data);
 
 	while (!freerdp_shall_disconnect(instance))
 	{
 		UINT32 tmp = freerdp_get_event_handles(instance->context, &handles[nCount],
 		                                       ARRAYSIZE(handles) - nCount);
 
-		if (tmp == 0)
+		if ((tmp == 0) || (nCount + tmp > ARRAYSIZE(handles)))
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply server/proxy/pf_client.c]: done

File: libfreerdp/core/freerdp.c [hunk 5]
```diff
@@ -660,26 +704,48 @@ BOOL freerdp_context_new(freerdp* instance)
...
+	metrics_free(ctx->metrics);
+	ctx->metrics = NULL;
+
+	CloseHandle(ctx->channelErrorEvent);
+	ctx->channelErrorEvent = NULL;
+
+	free(ctx->errorDescription);
+	ctx->errorDescription = NULL;
+
+	CloseHandle(ctx->abortEvent);
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/core/freerdp.c]: done

File: libfreerdp/core/freerdp.c [hunk 6]
```diff
@@ -660,26 +704,48 @@ BOOL freerdp_context_new(freerdp* instance)
...
+	graphics_free(ctx->graphics);
+	ctx->graphics = NULL;
+
+	PubSub_Free(ctx->pubSub);
+	ctx->pubSub = NULL;
+
+	metrics_free(ctx->metrics);
+	ctx->metrics = NULL;
+
+	CloseHandle(ctx->channelErrorEvent);
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/core/freerdp.c]: done

File: libfreerdp/core/freerdp.c [hunk 7]
```diff
@@ -115,19 +121,23 @@ BOOL freerdp_connect(freerdp* instance)
 
 	status = rdp_client_connect(rdp);
 
-	/* --authonly tests the connection without a UI */
-	if (instance->settings->AuthenticationOnly)
+	/* Pointers might have chanegd inbetween */
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/core/freerdp.c]: done

File: channels/rdpsnd/server/rdpsnd_main.c [hunk 8]
```diff
@@ -228,12 +228,11 @@ static UINT rdpsnd_server_recv_formats(RdpsndServerContext* context, wStream* s)
 
 static DWORD WINAPI rdpsnd_server_thread(LPVOID arg)
 {
-	DWORD nCount, status;
-	HANDLE events[8];
-	RdpsndServerContext* context;
+	DWORD nCount = 0, status;
+	HANDLE events[4] = { 0 };
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply channels/rdpsnd/server/rdpsnd_main.c]: done

**Comment by akallabeth** [comment]
@mfleisz as for the comments:

1. I´d keep the `freerdp_settings_*` calls unless they really decrease performance (don´t optimize before you have a problem ;))
2. Yes. we have an issue there, I´m on it but until resolved this works around the problem.

**Pull request merged**